src/app/cmd/flatten_layers.cpp

```diff
@@ -97,44 +154,94 @@ void FlattenLayers::onExecute()
+      const bool shrink = doc::algorithm::shrink_bounds(
+        image.get(), image->maskColor(), nullptr,
+        image->bounds(), bounds);
 
-      ImageRef cel_image;
+      // Skip when fully transparent
       Cel* cel = flatLayer->cel(frame);
+      if (!shrink) {
+        if (!newFlatLayer && cel)
+          executeAndAdd(new cmd::RemoveCel(cel));
+
+        continue;
+      }
+
+      // Apply shrunk bounds to new image
+      const ImageRef new_image(doc::crop_image(
+        image.get(), bounds, image->maskColor()));
+
+      // Replace image on existing cel
       if (cel) {
+
+        // TODO Keep cel links when possible
+
         if (cel->links())
           executeAndAdd(new cmd::UnlinkCel(cel));
 
-        cel_image = cel->imageRef();
+        const ImageRef cel_image = cel->imageRef();
         ASSERT(cel_image);
 
-        executeAndAdd(
-          new cmd::CopyRect(cel_image.get(), image.get(),
-                            gfx::Clip(0, 0, image->bounds())));
+        // Reset cel properties when flattening in-place
+        if (!newFlatLayer) {
+          executeAndAdd(new cmd::SetCelOpacity(cel, 255));
+          executeAndAdd(new cmd::SetCelPosition(cel,
+            area.src.x+bounds.x, area.src.y+bounds.y));
+        }
+
+        // Modify destination cel
+        executeAndAdd(new cmd::ReplaceImage(sprite, cel_image, new_image));
       }
+      // Add new cel on null
       else {
-        gfx::Rect bounds(image->bounds());
-        if (doc::algorithm::shrink_bounds(
-              image.get(), image->maskColor(), nullptr, bounds)) {
-          cel_image.reset(
-            doc::crop_image(image.get(), bounds, image->maskColor()));
-          cel = new Cel(frame, cel_image);
-          cel->setPosition(bounds.origin());
+        cel = new Cel(frame, new_image);
+        cel->setPosition(area.src.x+bounds.x, area.src.y+bounds.y);
+
+        // No need to undo adding this cel when flattening onto
+        // a new layer, as the layer itself would be destroyed,
+        // hence the lack of a command
+        if (newFlatLayer) {
           flatLayer->addCel(cel);
         }
+        else {
+          executeAndAdd(new cmd::AddCel(flatLayer, cel));
+        }
       }
     }
   }
 
+  // Notify observers when merging down
+  if ((m_options & Options::MergeDown) != 0)
+    doc->notifyLayerMergedDown(list.back(), flatLayer);
+
   // Add new flatten layer
-  if (newFlatLayer)
-    executeAndAdd(new cmd::AddLayer(list.front()->parent(), flatLayer, list.front()));
+  if (newFlatLayer) {
+    executeAndAdd(new cmd::AddLayer(
+      list.front()->parent(), flatLayer, list.front()));
+
+  }
+  // Reset layer properties when flattening in-place
+  else {
+    executeAndAdd(new cmd::SetLayerOpacity(flatLayer, 255));
+    executeAndAdd(new cmd::SetLayerBlendMode(
+      flatLayer, doc::BlendMode::NORMAL));
+    executeAndAdd(new cmd::SetLayerFlags(
+      flatLayer, doc::LayerFlags::Visible));
```

Comment:
Probably it'd be better to leave the flags as they were, with this PR the resulting destination layer is being locked.

Reply:
Right, I forgot to mention, it's possible to merge down when the destination layer isn't visible, which results in the source layer seemingly disappearing after merging as visibility remains disabled. The first time it happened I thought for a second that I had somehow deleted them, so I added this bit.

Another way to prevent such confusion would've been disabling from `MergeDownLayerCommand::onEnabled()` when the destination isn't visible, but honestly this is a very small detail, so it can be left as it was.

Reply:
You are right :thinking: Even worse, if the layer at the bottom is locked, the merge down command works anyway. We could leave this functionality as it's working now (leaving the flags as they were), and I'll create another issue about this.